File: app/src/main/java/com/example/drawingapp/loginscreen/RegisterActivity.java
package com.example.drawingapp.loginscreen;

// Android imports

import android.content.Intent;
import android.os.Bundle;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import com.basketball.logintestapplication.R;
import com.basketball.logintestapplication.model.User;
import com.basketball.logintestapplication.network.ApiService;
import com.basketball.logintestapplication.network.RetrofitClient;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * RegisterActivity
 * ----------------
 * This activity allows a new user to create an account by sending
 * a POST request with username and password to your Spring Boot API.
 */

public class RegisterActivity extends AppCompatActivity {

    // Declare UI components
    EditText etUsername, etPassword, etConfirmPassword; // Input fields
    Button btnRegister;                                 // Register button
    TextView tvGoToLogin;                               // Link to go back to login screen

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        // connect this activity to its layout file (res/layout/activity_register.xml)
        setContentView(R.layout.activity_register);

        // initialize UI components
        etUsername = findViewById(R.id.etUsername);
        etPassword = findViewById(R.id.etPassword);
        etConfirmPassword = findViewById(R.id.etConfirmPassword);
        btnRegister = findViewById(R.id.btnRegister);
        tvGoToLogin = findViewById(R.id.tvGoToLogin);

        // Create Retrofit API client
        ApiService api = RetrofitClient.getClient().create(ApiService.class);

        // When the user taps the "Register" button:
        btnRegister.setOnClickListener(v -> {
            // Read user input
            String username = etUsername.getText().toString();
            String password = etPassword.getText().toString();
            String confirm = etConfirmPassword.getText().toString();

            // Basic validation checks
            if (username.isEmpty() || password.isEmpty() || confirm.isEmpty()) {
                Toast.makeText(RegisterActivity.this, "Please fill all fields", Toast.LENGTH_SHORT).show();
                return;
            }

            if (!password.equals(confirm)) {
                Toast.makeText(RegisterActivity.this, "Passwords do not match", Toast.LENGTH_SHORT).show();
                return;
            }

            // Create request body for the API
            // Map<String, String> userData = new HashMap<>();
            // userData.put("username", username);
            // userData.put("password", password);


            // Send POST request to /api/users/register
            User user = new User();
            user.setUsername(username);
            user.setPassword(password);
            api.register(user).enqueue(new Callback<>() {
                @Override
                public void onResponse(@NonNull Call<User> call, @NonNull Response<User> response) {
                    if (response.isSuccessful()) {
                        // Registration successful - notify user
                        Toast.makeText(RegisterActivity.this, "Registration successful!", Toast.LENGTH_SHORT).show();

                        // Redirect back to LoginActivity
                        startActivity(new Intent(RegisterActivity.this, LoginActivity.class));
                        finish(); // Close this screen so the user can't go back with Back button
                    } else {
                        // API returned an error (e.g. username already taken)
                        Toast.makeText(RegisterActivity.this, "Registration failed", Toast.LENGTH_SHORT).show();
                    }
                }

                @Override
                public void onFailure(@NonNull Call<User> call, @NonNull Throwable t) {
                    // Connection/network failure
                    new AlertDialog.Builder(RegisterActivity.this)
                            .setTitle("Registration Error")          // Dialog title
                            .setMessage("Error: " + t.getMessage())  // Dialog message
                            .setPositiveButton("OK", null)           // OK button to dismiss
                            .show();
                }
            });
        });

        // When user taps "Already have an account? Log in"
        tvGoToLogin.setOnClickListener(v -> {
            Intent intent = new Intent(RegisterActivity.this, LoginActivity.class);
            startActivity(intent);
            finish();
        });
    }
}

File: app/src/main/java/com/example/drawingapp/loginscreen/RegisterActivity.kt
package com.example.drawingapp.loginscreen

// Android imports
import android.widget.Toast
import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.material3.ButtonDefaults
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.navigation.NavController
import com.example.drawingapp.R
import com.example.drawingapp.loginscreen.model.User
import com.example.drawingapp.loginscreen.network.ApiService
import com.example.drawingapp.loginscreen.network.RetrofitClient
import com.example.drawingapp.ui.whiteboardtheme.WhiteboardSimTheme
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import androidx.compose.material3.OutlinedTextField
/**
 * RegisterActivity
 * ----------------
 * This activity allows a new user to create an account by sending
 * a POST request with username and password to your Spring Boot API.
 */

@Composable
fun RegisterScreen(navCon: NavController) {
    WhiteboardSimTheme {
        // Declare UI components
        var username by remember { mutableStateOf("") }
        var password by remember { mutableStateOf("") }
        var confirmPassword by remember { mutableStateOf("") }
        val context = LocalContext.current

        val api = remember { RetrofitClient.getClient().create(ApiService::class.java) }

        // Background
        Box(
            modifier = Modifier.fillMaxSize()
        ) {
            Image(
                painter = painterResource(id = R.drawable.white_layer),
                contentDescription = "background",
                contentScale = ContentScale.FillBounds
            )
        }

        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(36.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Center

        ) {
            Image(
                painter = painterResource(id = R.drawable.logo),
                contentDescription = "QuickDraw Logo"
            )
            Spacer(Modifier.height(16.dp))

            Text(
                "Create a Free Account!",
                style = MaterialTheme.typography.headlineSmall,
                color = MaterialTheme.colorScheme.primary,
                fontWeight = FontWeight.Bold,
            )
            Spacer(Modifier.height(16.dp))

            OutlinedTextField(
                value = username,
                onValueChange = { username = it },
                label = { Text("Username") },
                singleLine = true
            )

            OutlinedTextField(
                value = password,
                onValueChange = { password = it },
                label = { Text("Password") },
                singleLine = true,
                visualTransformation = PasswordVisualTransformation()
            )

            OutlinedTextField(
                value = confirmPassword,
                onValueChange = { confirmPassword = it },
                label = { Text("Confirm Password") },
                singleLine = true,
                visualTransformation = PasswordVisualTransformation()
            )

            Spacer(Modifier.height(16.dp))

            Button(
                onClick = {
                    if (password != confirmPassword) {
                        Toast.makeText(context, "Passwords do not match", Toast.LENGTH_SHORT).show()
                        return@Button
                    }
                    val user = User(username = username, password = password)
                    api.register(user).enqueue(object : Callback<User> {
                        override fun onResponse(call: Call<User>, response: Response<User>) {
                            if (response.isSuccessful) {
                                Toast.makeText(context, "Registration Successful", Toast.LENGTH_SHORT)
                                    .show()
                                navCon.navigate("splash") {
                                    popUpTo("register") { inclusive = true }
                                }

                            } else {
                                Toast.makeText(context, "Registration failed", Toast.LENGTH_SHORT)
                                    .show()
                            }
                        }

                        override fun onFailure(call: Call<User>, t: Throwable) {
                            Toast.makeText(context, "Error: ${t.message}", Toast.LENGTH_LONG).show()
                        }

                    })
                },
                modifier = Modifier.fillMaxWidth(),
                colors = ButtonDefaults.buttonColors(
                    containerColor = MaterialTheme.colorScheme.primary,
                    contentColor = MaterialTheme.colorScheme.onPrimary,
                )
            ) {
                Text("Register")

            }
            Spacer(Modifier.height(12.dp))

            TextButton(onClick = { navCon.navigate("login") }) {
                Text(
                    "Already have an account? Log in",
                    color = MaterialTheme.colorScheme.secondary
                )
            }

        }
    }

}

@Preview
@Composable
fun RegisterScreenPreview() {
    RegisterScreen(navCon = NavController(LocalContext.current))
}
